initial import of opfcli command line tool

larsks opened the pull request:
This is my attempt at implementing what we talked about in
operate-first/SRE#299. It's an integrated command line tool written in
Go that currently provides equivalents for:

- onboarding.sh (called create-project)
- enable-monitoring.sh
- operate-first/apps#626 (called grant-access)

This is my first time writing Go, ever, so there may be lots of
non-idiomatic code. In your reviews, be kind, but be thorough smile.

larsks committed:
Initial import of opfcli command line tool

This is my attempt at implementing what we talked about in
operate-first/SRE#299. It's an integrated command line tool written in
Go that currently provides equivalents for:

- onboarding.sh (called create-project)
- enable-monitoring.sh
- operate-first/apps#626 (called grant-access)

This is my first time writing Go, ever, so there may be lots of
non-idiomatic code. In your reviews, be kind, but be thorough smile.

sesheta commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **durandom** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @durandom` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=operate-first%2Fopfcli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/operate-first/opfcli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["durandom"]} -->

sesheta commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **durandom** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @durandom` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=operate-first%2Fopfcli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/operate-first/opfcli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["durandom"]} -->

sesheta commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **durandom** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @durandom` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=operate-first%2Fopfcli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/operate-first/opfcli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["durandom"]} -->

larsks commented:
/cc @HumairAK 
/cc @tumido

4n4nd commented on README.md:
````diff
@@ -1,33 +1,189 @@
 
-## AI-CoE CI Github application
+```
+Grant a group acecss to a namespace.
````
*access

larsks committed:
Fix problems in pre-commit configuration

- Disable remove-tabs in pre-commit hook

  Standard go style includes using tabs for indentation. This is
  enforced by gofmt.

- Trailing whitespace in Markdown documents is significant

  We shouldn't be running the trailing-whitespace check on Markdown
  documents (because that's how you indicate a line break without
  introducing a new paragraph).

harshad16 commented:
/retest

harshad16 commented:
/test all

tumido commented on .pre-commit-config.yaml:
```diff
@@ -1,14 +1,14 @@
 repos:
-  - repo: git://github.com/Lucas-C/pre-commit-hooks
-    rev: v1.1.9
+  - repo: git://github.com/pre-commit/pre-commit-hooks
```
Nit, no need to have this repo imported twice. Just use the `pre-commit/pre-commit-hooks` hook id for `trailing-whitespace` as it is in the master.

larsks replied:
Yeah, not sure what I was thinking there.

tumido commented on README.md:
```diff
@@ -1,33 +1,189 @@
+(And will modify
+`cluster-scope/base/core/namespaces/project1/kustomization.yaml`)
+
+## License
```
There's already an GPL license besides readme (based on [ADR 0001](https://www.operate-first.cloud/blueprints/blueprint/docs/adr/0001-use-gpl3-as-license.md)). Do you want to claim copyright on this tool?

larsks replied:
According to "[How to use GNU licenses for your own software](https://www.gnu.org/licenses/gpl-howto.en.html)":

> This involves adding two elements to each source file of your program: a copyright notice (such as “Copyright 1999 Terry Jones”), and a statement of copying permission, saying that the program is distributed under the terms of the GNU General Public License (or the Lesser GPL, or the Affero GPL).

While I haven't gotten around to including the license in every source file yet, I think we do need to include the copyright notice and license information *somewhere*, and the README seems like a good choice.

larsks replied:
We can change it to be "Copyright ... The Operate First Team", but in order to use the GPL, we need to (a) assert copyright and (b) say that we're using it.

tumido commented on README.md:
```diff
@@ -1,33 +1,189 @@
-# Operate First template for repositories
+# opfcli
```
Can you please also include any dev docs or makefile in here? Like a base line for it - how to run `go build` and `go get`, etc.

larsks replied:
I've added a note about running `go build` to the README.

tumido reviewed (commented):
Looks great to me! :rocket: :+1:  Great baseline that we can build upon!

I have few docs nits only:
- license in README and copyright
- unnecessary changes in precommit config
- dev setup

HumairAK commented:
/retest

HumairAK commented on models/rolebinding.go:
```diff
@@ -0,0 +1,69 @@
+	if len(rolebinding.Subjects) == 0 {
+		rolebinding.Subjects = []Subject{sub}
+	} else {
+		rolebinding.Subjects = append(rolebinding.Subjects, sub)
```
afaik append works on nil/empty slices so not sure if this check is necessary

larsks replied:
I though I had run into a nilerror at one point, but you're right, it works without that.

HumairAK commented on models/resource.go:
```diff
@@ -0,0 +1,18 @@
+type kvmap map[string]string
+
+// Metadata represents the metadata section of a Kubernetes resource.
+type Metadata struct {
```
a code stylistic nit pick (feel free to disagree here) but I prefer it if the parent struct goes first, then children, so reorder it like: 
```golang
Resource struct { 
    ...
    Metadata Metadata
} 
// then define metadata struct

Metadata struct { 
    ....etc.
} 
```
it's easier to read imo, especially if the models get dense (same goes for the other models)

larsks replied:
I'm conditioned by decades of languages that fail if you try to refer to a data type before it's been defined. I mean, that fails in C; the following gives you `main.c:2:5: error: unknown type name ‘bar’`:

```
typedef struct {
    bar *Bar;
} foo;

typedef struct {
    int id;
} bar;
```

And it fails in Python, which gives you `NameError: name 'bar' is not defined` for this:

```
class foo(bar):
    def __init__(self):
        self.bar = bar()


class bar:
    ...
```

HumairAK replied:
I get that, if you feel strongly on this I won't push further, but it does seem to me that ~the standard for~ there is precedent for declaring struct representation of yaml/jsons seems to follow parent -> child type ordering, e.g. see how types are declared for argocd [here](https://github.com/argoproj/argo-cd/blob/master/pkg/apis/application/v1alpha1/types.go)

HumairAK commented on utils/fileutils.go:
```diff
@@ -0,0 +1,39 @@
+// PathIsDirectory returns true if the given path is a directory, and false if
+// it is not. If it cannot determine whether or not a path is a directory
+// (e.g., because of a permissions problem), it will log an error and exit.
+func PathIsDirectory(path string) (bool, error) {
```
This doesn't look like it's used anywhere?

larsks replied:
It's not anymore. It was at one point. I left it here on the theory that it might end up useful, but I'll remove it. We can always bring it back later.

HumairAK commented on models/resource.go:
```diff
@@ -0,0 +1,18 @@
+
+// Resource represents the core attributes of a Kubernetes resource. This
+// struct is embedded in all other Kubernetes resource models.
+type Resource struct {
```
Everytime I see this I do a double take, could we call this  `MetadataAPIVersionKind` ?

larsks replied:
Can you expand on why this makes you do a double take? I figured this was the basic information required to identify a Kubernetes resource, so the name seemed appropriate.

HumairAK replied:
My reasoning was: 

1. The entire resource is more than it's  MetadataAPIVersionKind (like `spec` for instance, or `users` for groups) 
2. For kustomizations, where `Resource` type is used, you also have `Resources` like [here](https://github.com/larsks/opfcli/blob/main/models/kustomization.go#L12), when I see something like that I think the plural would be a collection/list of the singular, but the 2 aren't really related. 

I don't care that much about this though, we can keep it as is if you prefer.

larsks replied:
Let's keep it for now; we can always rename it later if we come up with a compelling alternative.

HumairAK commented on models/kustomization.go:
```diff
@@ -0,0 +1,62 @@
+
+// Kustomization represents a kustomization file.
+type Kustomization struct {
+	Komponent  `yaml:",inline"`
```
IMO it's better to just list `Resource` / `Resources` fields here again, right now it seems like a `Komponent` _belongs_ in a `Kustomization`

HumairAK commented on cmd/main.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+func init() {
+	loglevel, err := strconv.Atoi(os.Getenv("OPF_LOGLEVEL"))
```
this env var doesn't like it's mentioned in the readme or anywhere else

larsks replied:
I've added a note to the README.

HumairAK commented on cmd/main.go:
```diff
@@ -0,0 +1,116 @@
+		&repoDirectory, "repodir", "R", "", "path to opf repository")
+
+	if err = config.BindPFlag("app-name", rootCmd.PersistentFlags().Lookup("app-name")); err != nil {
+		panic(err)
```
In the other parts of the code it seems the program exits via `log.Fatalf` any reason why we're doing a panic here instead of reporting a fatal error and exiting? (I'm not too versed on when we ought to panic over os.exit etc.)

larsks replied:
So, my though here was that this happens before logging is configured, and honestly this should *never* fail. Checking the error here is really just to make linters happy. Because a problem here represents a "how could this possible fail" situation, I figured a `panic` as appropriate.

HumairAK replied:
Ah that makes sense to me, thanks :+1:

HumairAK commented on cmd/api.go:
```diff
@@ -0,0 +1,200 @@
+	return nil
+}
+
+func createRoleBinding(projectName, groupName, roleName string) error {
```
`projectName` is unused

HumairAK commented on utils/kustomize.go:
```diff
@@ -0,0 +1,79 @@
+	log.Debugf("writing component kustomization for %s", path)
+	err := ioutil.WriteFile(
+		filepath.Join(path, "kustomization.yaml"),
+		[]byte(kustomOut), 0644,
```
I don't think you need to convert `kustomOut` here, `models.ToYAML` is already outputting a `[]byte`

larsks replied:
Good catch; artifact of previously returning `string`.

HumairAK commented:
The way the config file currently works doesn't seem very useful currently. But I guess we're still talking about it, so having the logic in place now will probably be useful later, so I say let's ignore that bit for now.

larsks commented:
> The way the config directory currently works doesn't seem very useful currently.

I thought the fact that it will load config from a top-level  configuration file was reasonable, although we don't really have much to configure at this point. I agree we can punt that discussion.

HumairAK commented:
/lgtm

HumairAK reviewed (approved):
Awesome work @larsks this is great!!

sesheta commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/operate-first/opfcli/pull/1#pullrequestreview-663613672" title="Approved">HumairAK</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=operate-first%2Fopfcli).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/operate-first/opfcli/blob/main/OWNERS)~~ [HumairAK]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
